diff --git a/dist/hooks/TransactionFeeHook.js b/dist/hooks/TransactionFeeHook.js
index ac07534fef9e1ab0031ca555ed2a7972c4cb3ce9..333d63f8636c0a1086129ebf6b7b70f6dbd5b532 100644
--- a/dist/hooks/TransactionFeeHook.js
+++ b/dist/hooks/TransactionFeeHook.js
@@ -45,9 +45,11 @@ export let TransactionFeeHook = class TransactionFeeHook extends ProvableTransac
         });
     }
     async start() {
-        this.persistedFeeAnalyzer = new RuntimeFeeAnalyzerService(this.runtime);
-        this.persistedFeeAnalyzer.config = this.config;
-        await this.persistedFeeAnalyzer.initializeFeeTree();
+        if (typeof window == 'undefined') {
+            this.persistedFeeAnalyzer = new RuntimeFeeAnalyzerService(this.runtime);
+            this.persistedFeeAnalyzer.config = this.config;
+            await this.persistedFeeAnalyzer.initializeFeeTree();
+        }
     }
     get config() {
         this.verifyConfig();
diff --git a/src/hooks/TransactionFeeHook.ts b/src/hooks/TransactionFeeHook.ts
index d22b2fc9dfe283480365c69d1ec16b472e5d5f71..2fd28c41f1131f2c8269f8f0c42766c0ac5d6761 100644
--- a/src/hooks/TransactionFeeHook.ts
+++ b/src/hooks/TransactionFeeHook.ts
@@ -77,10 +77,12 @@ export class TransactionFeeHook extends ProvableTransactionHook<TransactionFeeHo
   }
 
   public async start() {
-    this.persistedFeeAnalyzer = new RuntimeFeeAnalyzerService(this.runtime);
-    this.persistedFeeAnalyzer.config = this.config;
-    await this.persistedFeeAnalyzer.initializeFeeTree();
-  }
+    if (typeof window == 'undefined') {
+        this.persistedFeeAnalyzer = new RuntimeFeeAnalyzerService(this.runtime);
+        this.persistedFeeAnalyzer.config = this.config;
+        await this.persistedFeeAnalyzer.initializeFeeTree();
+    }
+}
 
   public get config() {
     this.verifyConfig();
